tests/worker/test_workflow.py

```diff
@@ -6990,3 +6994,242 @@ async def test_update_handler_semaphore_acquisition_respects_timeout(
         ),
     )
+
+
+deadlock_interruptible_completed = 0
+
+
+@workflow.defn(sandboxed=False)
+class DeadlockInterruptibleWorkflow:
+    @workflow.run
+    async def run(self) -> None:
+        # Infinite loop, which is interruptible via PyThreadState_SetAsyncExc
+        try:
+            while True:
+                pass
+        finally:
+            global deadlock_interruptible_completed
+            deadlock_interruptible_completed += 1
+
+
+async def test_workflow_deadlock_interruptible(client: Client):
+    # TODO(cretz): Improve this test and other deadlock/eviction tests by
+    # checking slot counts with Core. There are a couple of bugs where used slot
+    # counts are off by one and slots are released before eviction (see
+    # https://github.com/temporalio/sdk-core/issues/894).
+
+    # This worker used to not be able to shutdown because we hung evictions on
+    # deadlock
+    async with new_worker(client, DeadlockInterruptibleWorkflow) as worker:
+        # Start the workflow
+        assert deadlock_interruptible_completed == 0
+        handle = await client.start_workflow(
+            DeadlockInterruptibleWorkflow.run,
+            id=f"workflow-{uuid.uuid4()}",
+            task_queue=worker.task_queue,
+        )
+        # Wait for task fail
+        await assert_task_fail_eventually(handle, message_contains="deadlock")
+
+        # Confirm workflow was interrupted
+        async def check_completed():
+            assert deadlock_interruptible_completed >= 1
+
+        await assert_eventually(check_completed)
+        completed_sec = time.monotonic()
+    # Confirm worker shutdown didn't hang
+    assert time.monotonic() - completed_sec < 20
+
+
+deadlock_uninterruptible_event = threading.Event()
+deadlock_uninterruptible_completed = 0
+
+
+@workflow.defn(sandboxed=False)
+class DeadlockUninterruptibleWorkflow:
+    @workflow.run
+    async def run(self) -> None:
+        # Wait on event, which is not interruptible via PyThreadState_SetAsyncExc
+        try:
+            deadlock_uninterruptible_event.wait()
+        finally:
+            global deadlock_uninterruptible_completed
+            deadlock_uninterruptible_completed += 1
+
+
+async def test_workflow_deadlock_uninterruptible(client: Client):
+    # This worker used to not be able to shutdown because we hung evictions on
+    # deadlock
+    async with new_worker(client, DeadlockUninterruptibleWorkflow) as worker:
+        # Start the workflow
+        assert deadlock_uninterruptible_completed == 0
+        handle = await client.start_workflow(
+            DeadlockUninterruptibleWorkflow.run,
+            id=f"workflow-{uuid.uuid4()}",
+            task_queue=worker.task_queue,
+        )
+        # Wait for task fail
+        await assert_task_fail_eventually(handle, message_contains="deadlock")
+        # Confirm could not be interrupted
+        assert deadlock_uninterruptible_completed == 0
```

Comment:
Would it be more correct to assert that waiting for `>0` times out? Or is there a reason this line does not need to wait but similar checks in this and the `interruptible` test use `assert_eventually`?

Reply:
Hrmm, yes technically it could have been interrupted _after_ the task failed (though it'd have to be a quick server round trip to go faster than a Python interpreter thread). All of the eventual assertions are for positive assertions, we don't want timing out to be a valid thing that happens in successful tests usually (for test perf reasons, though sometimes we have to like deadlock itself). Can we think of another way to confirm that it cannot be interrupted? Would you settle for removing this assertion and having a whole new test that proves that `threading.Event`'s `wait` is uninterruptible via our interruption mechanism?

Reply:
I don't have a strong opinion. Would it help to put a mock object around `_handle_cache_eviction`? Perhaps having one task waiting for it to be called, confirming that we haven't hit the `finally`, and then setting the event. Something involving something like that?

Reply:
I have traditionally tried to avoid internal expectations on these tests and rather test behavior from the outside as a user may. I think it's probably ok to say that this assertion will just "usually" fail (i.e. fail like 99.9% of the time unless the server somehow does its round trip before this Python thread can execute its couple of lines) if Python stdlib changes behavior here instead of "always". It'll still be noticeable by us.

Reply:
Yes, that's OK with me too. Maybe add a comment explaining the technically possible thing.

Reply:
:+1: Didn't get to this pre-merge, may be able to in another PR